### DISCUSSION TITLE
Discussing high precision mouse mode on Linux

### DISCUSSION THREAD

**Discussion by yyny** [ppy/osu#13699] [Ideas]
Because the conversations keep getting locked.

@bdach

> > "High precision mouse" isn't very descriptive, and doesn't indicate that it grabs the mouse at all
>
> Where is this coming from, again? You keep stating this, but as far as I am aware high precision mouse/relative mouse mode does not grab the mouse. At most, what it does, is that it warps/recenters mouse position to keep the mouse inside the bounds of the screen. See SDL documentation.

See [the SDL source code](https://github.com/SpookyScaryDev/SDL2/blob/e53250eb69283a2714b3c229f5b66f9eb81e4e1b/src/events/SDL_mouse.c#L761) and [the resulting code path](https://github.com/SpookyScaryDev/SDL2/blob/e53250eb69283a2714b3c229f5b66f9eb81e4e1b/src/video/SDL_video.c#L2508). I don't understand why it matters how osu is grabbing the cursor, my window manager doesn't receive mouse events, so the cursor is being grabbed **by definition**. If all of you seemingly think that osu should not be doing this, then this is clearly unintended behaviour.

> Also as a side note this whole ordeal would probably not be anywhere near this painful if there was a video showing the behaviour you find undesirable.

I don't get a chance to, the conversations keep getting locked. Also, the undesirable behavior is the cursor grabbing, not the symptoms of it. I don't mind that my window manager doesn't receive cursor events during a cursor grab, ***that's the point of grabbing the cursor***. What I mind is that Osu is doing this when it doesn't have to, and seemingly unintentionally.

> If you still wanna discuss about your issue, you can post video about your issue at https://github.com/ppy/osu/discussions as already mentioned several times at ppy/osu#13684

That's not true. This is what @peppy said:

> If you'd like further help with using linux, please open a discussion instead.

And I don't need further help with using Linux at all, quite the opposite.

Also, once again, the symptoms are not the issue, and grabbing the pointer appears to cause problems on Windows and possibly OSX as well.

> It's still unclear to me what the actual problem is here.

Perhaps it is counterproductive then to lock issues before they are fixed or resolved.

@peppy 

> If there's an issue with the existing edge-of-window logic, that should be fixed. There should not be any OS-specific hard disables for functionality.

This is an issue with osu pretending to be a window manager. At this point, it seems like Osu shouldn't be doing edge-of-window detection *at all*, on *any* platform, unless it wants to switch screens. If Osu wants to lose focus when the cursor leaves the window, it should just not grab the cursor in the first place, so the window manager can handle edge detection by itself. The only time when it is acceptable to do this is during gameplay or when the window is fullscreen, as outlined in the MSDN article. It so happens to (partially) work for windows because DWM still receives cursor events when the cursor is grabbed.

> Closing as this completely breaks relative mouse mode and is not the correct direction.

> Please pay attention to the log overlay in the bottom left, wherein I have added log statements to denote when mayStayConfined is causing the relative mouse mode to be turned off (which is on every mouse movement, and every time the mouse returns to the screen).

> Even if (and that's a big if) dropping relative mouse mode was to be considered for linux it shouldn't be done in this roundabout way.

The provided patch does not drop relative mouse mode support at all, it only disables it while the user has not allowed Osu to confine (which remember, `grab = confine`) the cursor, because relative mouse mode grabs the cursor. If `Confine mouse cursor to window` is set to `Always` (or `During gameplay`), the option stays enabled. As I already outlined aswell, I do not think that is a good patch, but it gives the behaviour that appears to be currently intended:

> Again, to restate, this time using code, grab = confine mode.

# Video of the sympthoms

Notice how I try to resize, move, and bring focus to another window 3 times in a row, with and without "High precision mouse", which, as I've outlined many times by now, *grabs the cursor*.

https://user-images.githubusercontent.com/6576327/123843285-0915e280-d912-11eb-81af-dac0c31fc0fc.mp4

**Comment by bdach** [comment]
> See the SDL source code and the resulting code path.

Okay, my bad. I was not aware - the question in the PR thread was genuine.

Let me say this as an explanation for why this is the case: The reason that "high precision mouse" uses relative mouse mode is because it's the only cross-platform facility from SDL2 we can use where we can receive "high precision" input and control mouse sensitivity without going through xinput, etc.

Windows actually has a [platform-specific mouse handler](https://github.com/ppy/osu-framework/blob/99786238a02e0d6b69da86dd52e5506ee8ec0566/osu.Framework/Platform/Windows/WindowsMouseHandler.cs) that bypasses SDL2's relative mouse mode and uses the appropriate WinAPI endpoints (`WM_INPUT`). The idea of doing something for linux was thrown around a bit. If evdev or whatever can be used to provide something comparable to `WM_INPUT`, then that would be a great change in my view, with one caveat - it must be pure user-mode, like `WM_INPUT` is (so no udev rules, no world-`rw` input device streams, etc. - we've rode this rodeo with opentabletdriver and users are not all that understanding/happy about it).

> The provided patch does not drop relative mouse mode support at all, it only disables it while the user has not allowed Osu to confine

That is still not acceptable, though. Not without signalling to the user that the two settings are intertwined. If you consider hard-tying confine mode to relative mouse to be the way forward, then that should be the proposal, i.e. one setting should disable the other on a *config manager* level. Not by killing entire codepaths in SDL backend implementation internals with linux-specific checks.

Do note, however, that we've had the "high precision mouse" options disabled for a long time before on linux specifically (as it was completely busted), and had recurring requests to bring it back. Now that it's been around for a few good months, you're actually the *first* person raising concerns about this. So I wouldn't hold my breath that this is going to be treated as a high-priority item. Especially that you can just... turn off "high precision mouse", and use whichever settings xinput uses.

**Reply by yyny** [reply]
> Let me say this as an explanation for why this is the case: The reason that "high precision mouse" uses relative mouse mode is because it's the only cross-platform facility from SDL2 we can use where we can receive "high precision" input and control mouse sensitivity without going through xinput, etc.

Ok, so I checked this, `SDL_SetRelativeMouseMode(SDL_TRUE)` doesn't give any higher precision on the platforms I checked, it just reports raw (unaccelerated) mouse motion. With `xinput`, accelerated mouse (/joystick/trackpat/tablet etc.) motion is already sub-pixel precise anyway, and without `xinput` SDL will fall back to `XMotion` events, so either way, the option does nothing besides grabbing the cursor and using pre-accelerated motion data, neither of which provide any value for me. We should probably rename the option to `Raw mouse input` to reflect this, as other games do.

"high precision mouse" is misleading, as there are actually certain devices and drivers that can increase precision on demand, and some games are now doing this. I was under the impression that Osu might be doing something similar. As it appears, it does not.

And since the only _actual_ intention of calling `SDL_SetRelativeMouseMode` seems to be to get raw mouse motion:

> Windows actually has a platform-specific mouse handler that bypasses SDL2's relative mouse mode and uses the appropriate WinAPI endpoints (WM_INPUT). The idea of doing something for linux was thrown around a bit. If evdev or whatever can be used to provide something comparable to WM_INPUT, then that would be a great change in my view, with one caveat - it must be pure user-mode, like WM_INPUT is (so no udev rules, no world-rw input device streams, etc. - we've rode this rodeo with opentabletdriver and users are not all that understanding/happy about it).

```
~/SDL2$ rgrep SDL_EVDEV_Init                              
src/video/vivante/SDL_vivantevideo.c:    if (SDL_EVDEV_Init() < 0) {
src/video/kmsdrm/SDL_kmsdrmvideo.c:    SDL_EVDEV_Init();
src/video/raspberry/SDL_rpivideo.c:    if (SDL_EVDEV_Init() < 0) {
src/core/linux/SDL_evdev.h:extern int SDL_EVDEV_Init(void);
src/core/linux/SDL_evdev.c:SDL_EVDEV_Init(void)
```

`evdev` is extremely low level, SDL doesn't even enable it in user-mode environments.

`libinput` is cross-platform, as are X11 extensions like XInput/XI2.

After searching around a bit, [one of the first articles](https://keithp.com/blogs/Cursor_tracking/) I found explains exactly how to get raw mouse events with XInput, _without grabbing the cursor_.

These are actually the same events that SDL uses for `SetRelativeMouseMode`, and they can be easily retrieved from SDL using `SDL_EventState(SDL_SYSWMEVENT, SDL_ENABLE)`, [with some caviats](https://github.com/libsdl-org/SDL/blob/0dd7024d55019182b9e92d65221824a7a4dac992/src/video/x11/SDL_x11events.c#L312).

> That is still not acceptable, though. Not without signalling to the user that the two settings are intertwined. If you consider hard-tying confine mode to relative mouse to be the way forward, then that should be the proposal, i.e. one setting should disable the other on a config manager level. Not by killing entire codepaths in SDL backend implementation internals with linux-specific checks.

At the time, I was under the impression that that the behaviour on the side of the osu client was considered intended, since the issue got closed. I did attempt to suggest renaming the option.

**Reply by bdach** [reply]
> `SDL_SetRelativeMouseMode(SDL_TRUE)` doesn't give any higher precision on the platforms I checked, it just reports raw (unaccelerated) mouse motion

That is the point, yes. Otherwise the mouse accel slider would be just for show.

> We should probably rename the option to `Raw mouse input` to reflect this, as other games do.

It was named exactly that prior to #12006. I'm not honestly sure why it was changed, however.

> After searching around a bit, one of the first articles I found explains exactly how to get raw mouse events with XInput, without grabbing the cursor.

If you (or someone else) is willing to draft a proof-of-concept of libinput integration in framework then that may be something worth pursuing. Although I am still not sure how those raw libinput messages would interact with window focus. I imagine grabbing the mouse as SDL does could be a mechanism that prevents the game window from being clicked away from accidentally. Because if two windows are processing the same mouse events separately and getting different results/end mouse position, I imagine that for instance, the window manager can receive a mouse click that according to it should be outside of the game window and therefore take away the game window's focus, while the user was intending the movement to remain inside the window. Which I would hope you agree to be wrong/undesirable behaviour.

As to why we're not willing to spend time on it right here, right now: For now it's just not an item of high enough priority for the team to take on. We have 700 other issues here and another 300 in framework, and some of those actually concern missing game functionality for the platforms that the original osu! was already supporting. They take precedence over linux improvements.

**Reply by peppy** [reply]
This is all wrong though. Sensitivity support is nigh impossible without capturing the mouse. I'm not even going to waste words here.

It's not called "raw mouse" because that's a windows specific thing.

If the author wants to pursue further, please PR a fully working implementation after understanding all implications.

Locking this thread please do not open another.

@bdach and no, this is not "not supported because linux". the same behaviour is exhibited on windows. this whole thing is about a window overlapping the osu! window. this shouldn't be a thing in the first place. for users interested in such interactions, they should **not be using high precision mouse**. **turn it off and use the desktop mouse implementation, which will correctly handle these scenarios**.

**Comment by peppy** [comment]
Everything in your video is expected behaviour.

The one and only question I have: With confine mode turned off, if you move the mouse to the *edge of the osu! window* does it rescind mouse capture? If so, nothing is broken. If not, we may investigate further.

Note that that is the ONLY CASE which requires changes at our end. Anything further is your configuration choice.
